Spigot-Server-Patches/0379-Add-LivingEntity-getTargetEntity.patch

```diff
@@ -0,0 +1,184 @@
++        return grow(d0, d0, d0);
++    }
++    // Paper end
++
+     public AxisAlignedBB grow(double d0, double d1, double d2) {
+         double d3 = this.a - d0;
+         double d4 = this.b - d1;
+@@ -184,6 +191,7 @@ public class AxisAlignedBB {
+         return this.a < d3 && this.d > d0 && this.b < d4 && this.e > d1 && this.c < d5 && this.f > d2;
+     }
+ 
++    public boolean contains(Vec3D vec3d) { return b(vec3d); } // Paper - OBFHELPER
+     public boolean b(Vec3D vec3d) {
+         return this.e(vec3d.x, vec3d.y, vec3d.z);
+     }
+@@ -207,6 +215,7 @@ public class AxisAlignedBB {
+         return this.g(-d0);
+     }
+ 
++    public MovingObjectPosition calculateIntercept(Vec3D vec3d, Vec3D vec3d1) { return b(vec3d, vec3d1); } // Paper - OBFHELPER
+     @Nullable
+     public MovingObjectPosition b(Vec3D vec3d, Vec3D vec3d1) {
+         return this.a(vec3d, vec3d1, (BlockPosition)null);
+diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
+index f6c43bab4..a9688e08a 100644
+--- a/src/main/java/net/minecraft/server/Entity.java
++++ b/src/main/java/net/minecraft/server/Entity.java
+@@ -1518,6 +1518,7 @@ public abstract class Entity implements INamableTileEntity, ICommandListener, Ke
+         return new Vec3D((double) (f5 * f6), (double) (-f7), (double) (f4 * f6));
+     }
+ 
++    public Vec3D getEyePosition(float partialTicks) { return i(partialTicks); } // Paper - OBFHELPER
+     public Vec3D i(float f) {
+         if (f == 1.0F) {
+             return new Vec3D(this.locX, this.locY + (double) this.getHeadHeight(), this.locZ);
+@@ -2181,10 +2182,12 @@ public abstract class Entity implements INamableTileEntity, ICommandListener, Ke
+         return this.bP().size() < 1;
+     }
+ 
++    public float getCollisionBorderSize() { return aM(); } // Paper - OBFHELPER
+     public float aM() {
+         return 0.0F;
+     }
+ 
++    public Vec3D getLookVec() { return aN(); } // Paper - OBFHELPER
+     public Vec3D aN() {
+         return this.d(this.pitch, this.yaw);
+     }
+diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
+index 13dd55549..723df597c 100644
+--- a/src/main/java/net/minecraft/server/EntityLiving.java
++++ b/src/main/java/net/minecraft/server/EntityLiving.java
+@@ -2859,6 +2859,32 @@ public abstract class EntityLiving extends Entity {
+         return world.rayTrace(start, end, fluidCollisionOption);
+     }
+ 
++    public MovingObjectPosition getTargetEntity(int maxDistance) {
++        Vec3D start = getEyePosition(1.0F);
++        Vec3D look = getLookVec();
++        Vec3D end = start.add(look.x * maxDistance, look.y * maxDistance, look.z * maxDistance);
++        List<Entity> entityList = world.getEntities(this, getBoundingBox().expand(look.x * maxDistance, look.y * maxDistance, look.z * maxDistance).grow(1.0D, 1.0D, 1.0D), IEntitySelector.notSpectator().and(Entity::isInteractable));
```

Comment:
this is scary, as a large distance is going to make a large box.

could we maybe iterate every chunk in the line, iterate chunk entities, compare the distance as well as in factor if the angle of the players facing direction to the entities location actually is close to potentially being hit (like less than 3 degrees), and THEN do a ray trace to see if it gets hit.

that would significantly improve performance and avoid risk of a large entity lookup

Reply:
the code I used in async branch, in PlayerChunk.java, does some forward facing math stuff.

Reply:
The box approach isnt so bad. Went ahead and added a 120 distance limit (same as block raytrace). The box approach is used all over (projectile hit collisions, etc) and its the best we got.

`world#getEntities` doesnt return all entities in the chunks the box touches, there are upper and lower Y limits as the entities are stores in slices which are essentially 16 block high vertical slices in the chunks. Plus, it ignores chunks that are not loaded already.

This is a better approach than what plugins are currently implementing, which is a dot product comparision on all entities in a `nearby` api search (which includes entities out of view, including behind them).